hub/core/storage/ipfs.py

```diff
@@ -0,0 +1,483 @@
+            try:
+                self.links = self._get_links(self.cid)
+                self.ordered_links = self.get_hash(self.links, {})
+                cid = self.ordered_links[path]
+                res, content = self.gw.cat(cid)
+                b = bytes(content,'utf-8')
+                if res.status_code == 200:                
+                    return b
+                else:
+                    raise HTTPError (parse_error_message(res))
+            except StopIteration:
+                print(f"Path requested: {path}")
+            except FileNotFoundError:
+                print(f"Path requested: {path}, filenotfound")
+                raise KeyError(path)
+            except AssertionError:
+                print(f'Status code: {res.status_code}, Assertion error, content is {content}, path is {path}, cid query is {cid}, query is {query}, and in bytes we have {b}')
+            except TypeError:
+                print('Got type error.')
+        else:
+            raise KeyError(path)
+
+    def __setitem__(self, path, value):
+        """Sets the object present at the path with the value"""
+        if not self.stored:
+            _, res = self.gw.add_items(filepath=self.fpath, directory=True)
+            self.stored = True
+            self.get_set_cid(res)
+            return res
+        return True
+
+
+    def __delitem__(self):
+        """Remove pin of the object present at the CID.
+
+        Example:
+            ipfs_provider = IPFSProvider("https://ipfs.infura.io:5001/api/v0", cid="insert_cid")
+            del ipfs_provider
+
+        Args:
+            coreurl (str): gateway URL to access IPFS
+            cid (str): CID of the dataset stored on IPFS
+
+        Raises:
+            KeyError: If an incompatible gateway is provided. (removing pins is only available on local/Infura gateways)
+            HTTPError: If incorrect CID.
+        """
+        if self.coreurl != "https://ipfs.infura.io:5001/api/v0":
+            raise KeyError(self.coreurl)
+        res = self.gw.pin_rm(self.cid)
+        return res
+
+    def __iter__(self):
+        """Generator function that iterates over the keys of the provider.
+
+        Yields:
+            str: the path of the object that it is iterating over, relative to the root of the provider.
+        """
+        yield from self.cids
+
+    def __len__(self):
+        """Returns the number of files present inside the root of the provider.
+
+        Returns:
+            int: the number of files present inside the root.
+        """
+        return len(self.cids)
+
+    def _all_keys(self) -> Set[str]:
+        """Generator function that iterates over the keys of the provider.
+
+        Returns:
+            set: set of all keys present at the root of the provider.
+        """
+        if self.cids is None:
+            return []
+        return self.cids
+
+    def clear(self, prefix=""):
+        """Delete the contents of the provider."""
+        if self.cids:
```

Comment:
not accounting for prefix, also I don't see how setting self.cids to None will clear objects?